File: backend/app/services/dashboard_service.py
from sqlalchemy.orm import Session
from sqlalchemy import func

from app.models.node import GPUNode
from app.models.job import Job


def get_dashboard(db: Session):

    total_nodes = db.query(GPUNode).count()

    online_nodes = (
        db.query(GPUNode)
        .filter(GPUNode.status == "Online")
        .count()
    )

    offline_nodes = (
        db.query(GPUNode)
        .filter(GPUNode.status == "Offline")
        .count()
    )

    total_gpus = (
        db.query(func.sum(GPUNode.total_gpus))
        .scalar() or 0
    )

    available_gpus = (
        db.query(func.sum(GPUNode.available_gpus))
        .scalar() or 0
    )

    running_jobs = (
        db.query(Job)
        .filter(Job.status == "Running")
        .count()
    )

    pending_jobs = (
        db.query(Job)
        .filter(Job.status == "Pending")
        .count()
    )

    completed_jobs = (
        db.query(Job)
        .filter(Job.status == "Completed")
        .count()
    )

    return {
        "total_nodes": total_nodes,
        "online_nodes": online_nodes,
        "offline_nodes": offline_nodes,
        "total_gpus": total_gpus,
        "available_gpus": available_gpus,
        "used_gpus": total_gpus - available_gpus,
        "running_jobs": running_jobs,
        "pending_jobs": pending_jobs,
        "completed_jobs": completed_jobs,
    }

File: frontend/src/services/dashboardService.js
import API from "./api";

export const getDashboardStats = async () => {
  try {
    const [nodesResponse, jobsResponse] = await Promise.all([
      API.get("/nodes/"),
      API.get("/jobs/"),
    ]);

    const gpuNodes = nodesResponse.data;
    const jobList = jobsResponse.data;

    const totalNodes = gpuNodes.length;

    const onlineNodes = gpuNodes.filter(
      (node) => node.status === "Idle" || node.status === "Busy"
    ).length;

    const offlineNodes = gpuNodes.filter(
      (node) => node.status === "Offline"
    ).length;

    const totalGPUs = gpuNodes.reduce(
      (sum, node) => sum + node.total_gpus,
      0
    );

    const availableGPUs = gpuNodes.reduce(
      (sum, node) => sum + node.available_gpus,
      0
    );

    const usedGPUs = totalGPUs - availableGPUs;

    const utilization =
      totalGPUs > 0
        ? Math.round((usedGPUs / totalGPUs) * 100)
        : 0;

    const queuedJobs = jobList.filter(
      (job) => job.status === "Queued"
    ).length;

    const runningJobs = jobList.filter(
      (job) => job.status === "Running"
    ).length;

    const completedJobs = jobList.filter(
      (job) => job.status === "Completed"
    ).length;

    return {
      totalNodes,
      onlineNodes,
      offlineNodes,
      totalGPUs,
      availableGPUs,
      usedGPUs,
      utilization,
      queuedJobs,
      runningJobs,
      completedJobs,
    };
  } catch (error) {
    console.error("Failed to load dashboard data:", error);

    return {
      totalNodes: 0,
      onlineNodes: 0,
      offlineNodes: 0,
      totalGPUs: 0,
      availableGPUs: 0,
      usedGPUs: 0,
      utilization: 0,
      queuedJobs: 0,
      runningJobs: 0,
      completedJobs: 0,
    };
  }
};